app_history.md

```diff
@@ -0,0 +1,423 @@
+
+For single-page applications that want to update the current entry in the same manner as today's `history.replaceState()` APIs, we have the following:
+
+```js
+// Updates the URL shown in the address bar, as well as the url property.
+appHistory.updateCurrentEntry({ url });
+
+// Only updates the state property.
+appHistory.updateCurrentEntry({ state });
+
+// Update both at once.
+appHistory.updateCurrentEntry({ url, state });
+```
+
+_TODO: more realistic example, maybe something Redux-esque like `appHistory.updateCurrentEntry({ state: {...appHistory.currentEntry.state, newKey: newValue } })`._
+
+Similarly, to push a new entry in the same manner as today's `history.pushState()`, we have the following:
+
+```js
+// Pushes a new entry onto the app history list, copying the URL and state from the current one.
+await appHistory.pushNewEntry();
+
+// Copy over the URL, but use new state:
+await appHistory.pushNewEntry({ state });
+
+// Copy over the state, but use a new URL:
+await appHistory.pushNewEntry({ url });
+
+// Don't copy over anything:
+await appHistory.pushNewEntry({ url, state });
+```
+
+As with `history.pushState()` and `history.replaceState()`, the new URL here must be same-origin and only differ in the path, query, or fragment portions from the current document's current URL. And as with those, you can use relative URLs.
+
+Note that `appHistory.pushNewEntry()` is asynchronous. As with other [navigations through the app history list](#navigation-through-the-app-history-list), pushing a new entry can be intercepted or canceled, so it will always be delayed at least one microtask.
+
+_TODO: need a realistic example of using `appHistory.pushNewEntry()` instead of `navigationEvent.respondWith()`._
+
+Crucially, `appHistory.currentEntry` stays the same regardless of what iframe navigations happen. It only reflects the current entry for the current frame. The complete list of ways the current app history entry can change are:
+
+- Via the above APIs, used by single-page apps to manage their own history entries.
+
+- A fragment navigation, which will act as `appHistory.pushNewEntry({ url: urlWithFragment })`, i.e. it will copy over the state.
+
+- A full-page navigation to a different document. This could be an existing document in the browser's back/forward cache, or a new document. In the latter case, this will generate a new entry on the new page's `window.appHistory` object, somewhat similar to `appHistory.pushNewEntry({ url: navigatedToURL, state: null })`. Note that if the navigation is cross-origin, then we'll end up in a separate app history list for that other origin.
+
+### Inspection of the app history list
+
+In addition to the current entry, the entire list of app history entries can be inspected, using `appHistory.entries()`, which returns an array of `AppHistoryEntry` instances. (Recall that all app history entries are same-origin and for the current frame, so this is not a security issue.)
```

Comment:
Would this be same-origin-contiguous? As in

1. `example.com/foo`
2. `example.com/bar`
3. `jakearchibald.com/whatever`
4. `example.com/hello`

If you are on session history item 4, will `appHistory.entries()` give you details about 1 and 2?

Reply:
Definitely contiguous! Will add.